jlm/rvsdg/theta.hpp

```diff
@@ -223,278 +180,187 @@ class ThetaNode final : public StructuralNode
+  MapPreLoopVar(const rvsdg::output & argument) const;
 
-/**
- * Represents the predicate result of a theta subregion.
- */
-class ThetaPredicateResult final : public RegionResult
-{
-  friend ThetaNode;
+  /**
+   * \brief Maps variable at end of loop iteration to full varibale description.
+   *
+   * \param result
+   *   Result of theta region.
+   *
+   * \returns
+   *   The loop variable description.
+   *
+   * \pre
+   *   \p result must be a result to the subregion of this node.
+   *
+   * Returns the full description of the loop variable corresponding
+   * to this variable at the end of each loop iteration.
+   */
+  [[nodiscard]] LoopVar
+  MapPostLoopVar(const rvsdg::input & result) const;
 
-public:
-  ~ThetaPredicateResult() noexcept override;
+  /**
+   * \brief Maps variable at exit to full varibale description.
+   *
+   * \param output
+   *   Output of this theta node
+   *
+   * \returns
+   *   The loop variable description
+   *
+   * \pre
+   *   \p output must be an output of this node
+   *
+   * Returns the full description of the loop variable corresponding
+   * to this loop exit value.
+   */
+  [[nodiscard]] LoopVar
+  MapOutputLoopVar(const rvsdg::output & output) const;
 
-  ThetaPredicateResult &
-  Copy(rvsdg::output & origin, StructuralOutput * output) override;
+  /**
+   * \brief Returns all loop variables.
+   *
+   * \returns
+   *   List of loop variable descriptions.
+   */
+  [[nodiscard]] std::vector<LoopVar>
+  GetLoopVars() const;
 
 private:
-  explicit ThetaPredicateResult(rvsdg::output & origin)
-      : RegionResult(origin.region(), &origin, nullptr, ControlType::Create(2))
-  {
-    JLM_ASSERT(is<ThetaOperation>(origin.region()->node()));
-  }
-
-  static ThetaPredicateResult &
-  Create(rvsdg::output & origin)
-  {
-    auto thetaResult = new ThetaPredicateResult(origin);
-    origin.region()->append_result(thetaResult);
-    return *thetaResult;
-  }
+  // Calling RemoveThetaInputsWhere/RemoveThetaOutputsWhere can result
+  // in inputs (and pre-loop arguments) and outputs (and post-loop results)
+  // to become unmatched. In this case, the theta node itself has
+  // "invalid" shape until fixed properly.
+  // The indices of unmatched inputs/outputs are tracked here to
+  // detect this situation, and also to provide correct mapping.
+  // Computing the mapping is a bit fiddly as it requires adjusting
+  // indices accordingly, should seriously consider whether this
+  // is really necessary or things can rather be reformulated such that
+  // inputs/outputs are always consistent.
+
+  std::optional<std::size_t>
+  MapInputToOutputIndex(std::size_t index) const noexcept;
+
+  std::optional<std::size_t>
+  MapOutputToInputIndex(std::size_t index) const noexcept;
+
+  void
+  MarkInputIndexErased(std::size_t index) noexcept;
+
+  void
+  MarkOutputIndexErased(std::size_t index) noexcept;
+
+  std::vector<std::size_t> unmatchedInputs;
+  std::vector<std::size_t> unmatchedOutputs;
```

Comment:
Is this really necessary? The only places where theta variables are removed should be DNE, and there we use the `RemoveThetaInputsWhere` and `RemoveThetaOutputsWhere` immediately after each other to ensure that the inconsistent state is very local. Where else are theta inputs/outputs are removed such that this becomes necessary?